base/experiments/ad9910_ram.py

```diff
@@ -0,0 +1,206 @@
+            case RAMType.PHASE:
+                self.dest = RAM_DEST_POW
+                dds.turns_to_ram(data, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+                self.asf = dds.amplitude_to_asf(base_amplitude)
+            case RAMType.AMP:
+                self.dest = RAM_DEST_ASF
+                dds.amplitude_to_ram(data, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+                self.pow = dds.turns_to_pow(base_phase)
+            case RAMType.POLAR:
+                self.dest = RAM_DEST_POWASF
+                phase, amp = data
+                dds.turns_amplitude_to_ram(phase, amp, ram)
+                self.ftw = dds.frequency_to_ftw(base_frequency)
+            case _:
+                raise NotImplementedError
+
+        self.start_addr = 0
+        self.end_addr = len(ram) - 1    # Inclusive
+
+        # Configure ther use of OSK. Note that using different OSK settings for
+        # different channels introduces step swiching latencies
+        if ram_type == RAMType.AMP or ram_type == RAMType.POLAR:
+            # Disable OSK. OSK amplitude as it has a higher priority than RAM.
+            self.osk_enable = 0
+        else:
+            # Enable OSK. There are no other utilizable amplitude data source.
+            self.osk_enable = 1
+
+        # Conversion to list to avoid type mismatch in the kernel
+        self.ram = ram.tolist()
+
+        # Note: Integer conversion may cause inaccuracy
+        self.step = int(ramp_interval * dds.sysclk / 4.0)
+
+        self.ramp_mode = ramp_mode
+
+
+class RAMProfileMap:
+    def __init__(self, core):
+        """Initialize a RAM profile to DDS mapping
+
+        Args:
+            core: The ARTIQ Core instance for time control.
+        """
+        self.ram_profile_map = []
+        self.cplds = []
+        self.core = core
+
+    def append(self, dds, ram_profile):
+        """Append a RAM profile into the builder. In addition, register the
+            CPLDs that have DDSes that playback the RAM profile.
+
+        Args:
+            dds: The DDS that will playback the RAM profile.
+            ram_profile: The RAM profile.
+        """
+        self.ram_profile_map.append((dds, ram_profile))
+
+        # Add the associlated CPLD to the list if not already there.
+        # ARTIQ-Python does not support iterration of a Python builtin set
+        # The workaround is to maintain a list, but convert it into set to
+        # avoid CPLD duplications.
+        cplds = set(self.cplds)
+        cplds.add(dds.cpld)
+        self.cplds = list(cplds)
+
+    @kernel
+    def program(self):
+        """Program the RAM profiles to the corresponding AD9910s.
+
+        """
+        # Switch to profile 0 for RAM. This is the default profile.
+        # Using other profiles for RAM is possible with appropriate parameters
+        # for future AD9910 function calls.
+        for cpld in self.cplds:
+            cpld.set_profile(0)
```

Comment:
Let's discuss about profiles and how we should use them in the code. I think I set the default DDS profile to 7 for typical use (single tone operation). Therefore, profile 0 for RAM modulation could be great. I am just not familiar with how this works - after setting the RAM modulation data in profile 0, can I go back to profile 7 to run at the single tone parameters, and then go to profile 0 to run the RAM modulation data?

Reply:
I think you can. This is probably the flow:

- Program the RAM & RAM profiles.
- Unset `ram_enable` in CFR1 for every ad9910 in question.
- `cpld.set_profile(7)` for every Urukul in question.
- Go back to RAM by undoing the previous 2 steps.

Reply:
Okay I understand it better now. I think the `RAMProfileMap.program` should likely be separated into several smaller functions for each of the step here. At the beginning of `run`, users can program the RAM sequence needed using `load_ram_data`. When the user wants to switch the ad9910 from single tone to RAM modulation, they can call `enter_ram_mode`. Then the user can call `run_modulation_sequence`, and finally `exit_ram_mode`.

I don't know if the above makes completely sense or if I still miss some steps.

Reply:
I've renamed `program()` to `load_ram()`. We do have to load the profile when loading the RAM content.
And `enable()`, `commit_enable()`, `disable()`, `commit_disable()`.
The commit methods changes profile from single-tone (7) to RAM (0). It already triggers register content update, so no need to pulse I/O update anymore.